src/sage/interfaces/maxima_abstract.py

```diff
@@ -1679,8 +1702,12 @@ def __iter__(self):
             sage: [e._sage_() for e in L]
             [0, x, 2*x^2, 3*x^3, 4*x^4, 5*x^5]
         """
-        for i in range(len(self)):
-            yield self[i]
+        # Maxima lists are linked lists with O(n) random access time
+        # hence we create a copy and pop elements from the front instead
+        # so that __iter__ has complexity as O(n) and not O(n^2)
+        copied_list = self.copylist()
+        for i in range(len(copied_list)):
```

Comment:
does `self.first()`/`self.rest()` work? I think it should also be O(n) but avoid the overhead of initial copy. (also in islice case you avoid the initial copy entirely)